Fix: use hatch for workflow

lwasser opened the pull request:
This is what the intro looks like fully committing to hatch. Setting up that config.toml file for default hatch settings is important as a starting point. 

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206595156181300

lwasser commented:
@all-contributors please add @JacksonBurns for code, review, tutorials

lwasser commented on tutorials/1-installable-code.md:
```diff
@@ -295,44 +275,141 @@ For instance, a `build-system` table most often holds two (2) variables:
 [next lesson when you add additional metadata / information to this file.](5-pyproject-toml.md)
 :::
 
+- Open up the `pyproject.toml` file that Hatch created in your favorite text editor. Is should look something like the example below.
```
fixed! great catch!

lwasser commented on tutorials/1-installable-code.md:
```diff
@@ -158,56 +150,43 @@ Neither 'setup.py' nor 'pyproject.toml' found.
-Inside the project directory:
+* Open your shell or preferred terminal.
+* cd to the location where you'd like your package directory to live. Hatch will create the package directory for you
+* Choose a name for your package, preferably in lowercase and without spaces. For this tutorial we'll use `pyospackage`.
```
excellent catch. i've changed this to:

* Open your shell or preferred terminal.
* Use the shell `cd` command to navigate in your shell to the location where you'd like your package directory to live. Hatch will create the package directory for you
* Choose a name for your package. The name should:
    * Have no spaces (*Required*)
    * Use all lowercase characters (*Recommended*). For this tutorial we will use `pyospackage`.

lwasser commented on package-structure-code/python-package-build-tools.md:
```diff
@@ -345,14 +347,14 @@ as building your documentation locally. This means that you could potentially dr
 Version bumping| ✅ | Hatch supports you bumping the version of your package using standard semantic version terms patch; minor; major
 Follows current packaging standards|✅|Hatch supports current packaging standards for adding metadata to the **pyproject.toml** file.
-Install your package in editable mode|✖✅| You can install your package in editable mode using `pip install -e .` Hatch mentions [editable installs](https://hatch.pypa.io/latest/config/build/#dev-mode) but refers to pip in its documentation.
+Install your package in editable mode|✅| Hatch will install your package into any of it's environments by default in editable mode. You can install your package in editable mode manually using `pip install -e .` Hatch mentions [editable installs](https://hatch.pypa.io/latest/config/build/#dev-mode) but refers to pip in its documentation.
```
i actually just discovered that the caption for the graphic also has a note to myself that no one caught! i'm going to pull this page out and merge it separately with all of these typo fixes!

ucodery commented on tutorials/get-to-know-hatch.md:
```diff
@@ -0,0 +1,128 @@
+[instructions here](https://hatch.pypa.io/latest/install/).
+
+:::{tip}
+If you are comfortable using [pipx](https://pipx.pypa.io/stable/) to install hatch, we encourage you to do so. pipx will ensure that your package is available across all of your Python environments on your computer rather than just in the environment that you install it into.
+
+However if you are not sure about pipx, you can install hatch using pip or conda.
```
I love this new pipx tip!

lwasser replied:
yay! pipx is really growing on me. i still don't fully understand how it works on the back end but as a user, having hatch and nox, etc available to me in "all the pythons" 😆 , is really nice.

ucodery commented on tutorials/get-to-know-hatch.md:
```diff
@@ -0,0 +1,128 @@
+need to use this feature in this beginner friendly tutorial series
+but we will introduce it in later tutorials.
+
+Your config.toml file should look something like the one below. It's ok if you have
```
incomplete sentence

lwasser replied:
eeks. fixed. i have no idea what i was trying to write there. great catch!

lwasser commented on tutorials/1-installable-code.md:
```diff
@@ -158,56 +150,45 @@ Neither 'setup.py' nor 'pyproject.toml' found.
+* Use the shell `cd` command to navigate in your shell to the location where you'd like your package directory to live. Hatch will create the package directory for you
+* Choose a name for your package. The name should:
+    * Have no spaces (*Required*)
+    * Use all lowercase characters (*Recommended*). For this tutorial we will use `pyospackage`.
```
oh @ucodery gosh i had no idea that hatch normalized file names. 

i just ran some tests too:

```
❯ hatch new test-me-foreee
test-me-foreee
├── src
│   └── test_me_foreee
│       ├── __about__.py
│       └── __init__.py
├── LICENSE.txt
├── README.md
└── pyproject.toml
```
so its rules are

* dashes for the top level directory
* dashes for the project name in the pyproject.toml
* underscores for the package directory name 

I wonder why it chose underscores only for that directory? that always confused me when i was newer to packaging - where the final name is pulled from and how the final package name could be different from the directory name.

lwasser replied:
Ok this may be a bit too much. but i suspect the project directory name thing will trip users up. Mostly because it did so for me for a long time. 

<img width="853" alt="Screenshot 2024-02-14 at 6 29 20 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/b5e2735e-36c1-4c50-8524-48594927814f">

ucodery replied:
Personally, I think a call out that hatch makes some decisions for you is enough. having that much example output is just a wall of text, especially for those that might not care about the admonition.

> I wonder why it chose underscores only for that directory? 

This is the one name choice that is not really a choice. The package directory becomes the final python package, the thing that gets imported, and so must follow python variable naming rules.

I suspect that hatch is running the standard [name normalization](https://packaging.python.org/en/latest/specifications/name-normalization/#name-format) on the `new` name, using that for the package directory, then using a `.replace('_', '-')` for all other names it defines.

lwasser commented on tutorials/1-installable-code.md:
```diff
@@ -295,44 +275,141 @@ For instance, a `build-system` table most often holds two (2) variables:
+
+:::{note}
+You will learn how to automate defining a package
+version using git tags in the version and release your package lesson.
```
great suggestion. the actual tutorial on doing this with hatch isn't created ... yet! maybe i'll make this a TODO so it's hidden until we actually build and publish that lesson. that might be a good middle ground for now!

lwasser commented:
Ok y'all. this PR has been open for a few weeks and we haven't gotten a tremendous amount of additional feedback! thank you so much @ucodery and @JacksonBurns for the feedback! i think i've addressed all of the comments! I'm going to go ahead and merge it.
